codec/src/main/java/io/netty/handler/codec/compression/BrotliDecoder.java

```diff
@@ -0,0 +1,164 @@
+
+    private enum State {
+        DONE, NEEDS_MORE_INPUT, ERROR
+    }
+
+    static {
+        try {
+            Brotli.ensureAvailability();
+        } catch (Throwable throwable) {
+            throw new ExceptionInInitializerError(throwable);
+        }
+    }
+
+    private final int inputBufferSize;
+    private DecoderJNI.Wrapper decoder;
+    private boolean destroyed;
+
+    public BrotliDecoder() {
+        this(8 * 1024);
+    }
+
+    public BrotliDecoder(int inputBufferSize) {
+        this.inputBufferSize = inputBufferSize;
+    }
+
+    private ByteBuf pull(ByteBufAllocator alloc) {
+        ByteBuffer nativeBuffer = decoder.pull();
+        ByteBuf copy = alloc.buffer(nativeBuffer.remaining());
+        copy.writeBytes(nativeBuffer);
+        return copy;
+    }
+
+    private State decompress(ByteBuf input, List<Object> output, ByteBufAllocator alloc) {
+        for (;;) {
+            switch (decoder.getStatus()) {
+                case DONE:
+                    return State.DONE;
+
+                case OK:
+                    decoder.push(0);
+                    break;
+
+                case NEEDS_MORE_INPUT:
+                    if (decoder.hasOutput()) {
+                        output.add(pull(alloc));
+                    }
+
+                    if (!input.isReadable()) {
+                        return State.NEEDS_MORE_INPUT;
+                    }
+
+                    ByteBuffer decoderInputBuffer = decoder.getInputBuffer();
+                    decoderInputBuffer.clear();
+                    int readBytes = readBytes(input, decoderInputBuffer);
+                    decoder.push(readBytes);
+                    break;
+
+                case NEEDS_MORE_OUTPUT:
+                    output.add(pull(alloc));
+                    break;
+
+                default:
+                    return State.ERROR;
+            }
+        }
+    }
+
+    private static int readBytes(ByteBuf in, ByteBuffer dest) {
+        int limit = Math.min(in.readableBytes(), dest.remaining());
+        ByteBuffer slice = dest.slice();
+        slice.limit(limit);
+        in.readBytes(slice);
+        dest.position(dest.position() + limit);
+        return limit;
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        decoder = new DecoderJNI.Wrapper(inputBufferSize);
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+        if (destroyed) {
+            // Skip data received after finished.
+            in.skipBytes(in.readableBytes());
+            return;
+        }
+
+        if (!in.isReadable()) {
+            return;
+        }
+
+        try {
+            State state = decompress(in, out, ctx.alloc());
+            if (state == State.DONE) {
+                destroy();
+            } else if (state == State.ERROR) {
+                throw new DecompressionException("Brotli stream corrupted");
```

Comment:
Can brotli recover from errors? If not, consider invoking `destroy()` method before throwing an exception.

Reply:
done